### PR TITLE
Add diabetes downloads

### PR DESCRIPTION
**Pull request by govindkrjoshi**
**Story card:** [sc-7794](https://app.shortcut.com/simpledotorg/story/7794/monthly-district-data-download)
**Story card:** [sc-7796](https://app.shortcut.com/simpledotorg/story/7796/monthly-facility-data-download)

## Because

We need to report diabetes monthly district data and monthly facility data in CSV format

## This addresses

Make diabetes monthly district data and monthly facility data downloadable in CSV format

## Test instructions

- Make sure the Flipper flag named diabetes_management_reports is enabled
- Go to Report > Diabetes > Downloads(DM)
- You should be able to see the reports listed in the menu
- Click it to download and open to see the reports

### CODE REVIEW

File: app/components/dashboard/diabetes/downloads_dropdown_component.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  def initialize(region:, period:, report_scope:, current_admin:)
+    @region = region
+    @period = period
+    @report_scope = report_scope
+    @current_admin = current_admin
+  end
+
+  def monthly_state_data_path
+    reports_region_diabetes_monthly_state_data_path(
+      @region,
```
Comment (govindkrjoshi) [inline app/components/dashboard/diabetes/downloads_dropdown_component.rb:15]: [Style] Can use `region`

File: app/components/dashboard/hypertension/downloads_dropdown_component.html.erb [hunk 2]
```diff
@@ -3,41 +3,42 @@
...
 
-    <%= link_to(reports_region_download_path(@region, report_scope: params[:report_scope], period: :quarter, format: :csv), class: "dropdown-item") do %>
+    <%= link_to(region_download_path, class: "dropdown-item") do %>
       <i class="far fa-file w-16px ta-center mr-8px c-blue"></i>
       Quarterly cohort report
     <% end %>
 
-    <% if @region.district_region? && current_admin.feature_enabled?(:monthly_district_report)%>
-      <%= link_to(reports_region_monthly_district_report_path(@region, report_scope: params[:report_scope], period: params.dig(:period, :value), format: :zip), class: "dropdown-item") do %>
+    <% if region.district_region? && current_admin.feature_enabled?(:monthly_district_report)%>
```
Comment (govindkrjoshi) [inline app/components/dashboard/hypertension/downloads_dropdown_component.html.erb:18]: Since we are using components, you can move this condition into a methion, maybe something like `show_monthly_district_report?`

This way, we can keep our view clean and easy to read.

Same of other conditionals.

File: app/components/dashboard/hypertension/downloads_dropdown_component.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+  def initialize(region:, report_scope:, period:, current_admin:)
+    @region = region
+    @period = period
+    @report_scope = report_scope
+    @current_admin = current_admin
+  end
+
+  def region_download_path
+    reports_region_download_path(@region, report_scope: report_scope, period: :quarter, format: :csv)
```
Comment (govindkrjoshi) [inline app/components/dashboard/hypertension/downloads_dropdown_component.rb:14]: [Style] As that from [here](https://github.com/simpledotorg/simple-server/pull/4114/files#diff-cd1f09bfbd003f015175838bd04308ce20e234e5f52e0d17c72b709a7a132ca0R15 )

File: app/controllers/reports/regions_controller.rb [hunk 4]
```diff
@@ -272,22 +280,70 @@ def monthly_state_data_report
...
+    csv = MonthlyStateData::Exporter.new(
+      exporter: MonthlyStateData::Diabetes.new(
+        region: @region,
+        period: @period,
+        medications_dispensation_enabled: @medications_dispensation_enabled
+      )
+    ).report
     report_date = @period.to_s.downcase
-    filename = "monthly-district-report-#{@region.slug}-#{report_date}.zip"
+    filename = "monthly-district-data-#{@region.slug}-#{report_date}.csv"
```
Comment (govindkrjoshi) [inline app/controllers/reports/regions_controller.rb:312]: Maybe we should rename these file names for diabetes and potentially hypertension?

File: app/models/reports/region_summary_schema.rb [hunk 5]
```diff
@@ -350,9 +354,9 @@ def visited_without_bp_taken(with_ltfu: false)
 
     memoize def diabetes_patients_with_bs_taken
```
Comment (govindkrjoshi) [inline app/models/reports/region_summary_schema.rb:355]: [Naming] Let's prefix the methods also with `adjusted_`?

File: spec/controllers/reports/regions_controller_spec.rb [hunk 6]
```diff
@@ -459,7 +459,30 @@
       files = []
       Timecop.freeze("June 1 2020") do
         sign_in(cvho.email_authentication)
-        get :monthly_district_report, params: {id: district.slug, report_scope: "district", format: "zip"}
+        get :hypertension_monthly_district_report, params: {id: district.slug, report_scope: "district", format: "zip"}
+      end
+      expect(response).to be_successful
+
+      Zip::File.open_buffer(response.body) do |zip|
```
Comment (govindkrjoshi) [inline spec/controllers/reports/regions_controller_spec.rb:466]: Good test :clap:

File: spec/controllers/reports/regions_controller_spec.rb [hunk 7]
```diff
@@ -550,37 +573,163 @@
...
-      expect(MonthlyDistrictDataService).to receive(:new).with(region, period, medications_dispensation_enabled: false).and_call_original
-      get :monthly_district_data_report,
-        params: {id: region.slug, report_scope: "district", format: "csv", period: period.value}
+
+      exporter = MonthlyDistrictData::Hypertension.new(
+        region: region,
+        period: period,
+        medications_dispensation_enabled: false
+      )
+      allow(MonthlyDistrictData::Hypertension).to receive(:new).and_return(exporter)
```
Comment (govindkrjoshi) [inline spec/controllers/reports/regions_controller_spec.rb:582]: [Suggestion] A comment about why you need to mock `:new` would be helpful when returning to this test in the future.

File: spec/controllers/reports/regions_controller_spec.rb [hunk 8]
```diff
@@ -550,37 +573,163 @@
...
   end
 
-  describe "#monthly_state_data_report" do
+  describe "#diabetes_monthly_district_data" do
+    let(:facility_group) { create(:facility_group, organization: organization) }
+    let(:facility) { create(:facility, facility_group: facility_group) }
+    let(:region) { facility.region.district_region }
+
+    it "returns 401 when user is not authorized" do
+      facility
```
Comment (govindkrjoshi) [inline spec/controllers/reports/regions_controller_spec.rb:595]: Instead of  forcing the facility to be created like this, since you don't need a reference to the created object, you can create the objects in a before block instead of using lets.

File: spec/controllers/reports/regions_controller_spec.rb [hunk 9]
```diff
@@ -550,37 +573,163 @@
...
+      period = Period.month("July 2018")
+      exporter = MonthlyDistrictData::Diabetes.new(
+        region: region,
+        period: period,
+        medications_dispensation_enabled: false
+      )
+      allow(MonthlyDistrictData::Diabetes).to receive(:new).and_return(exporter)
+      expect(MonthlyDistrictData::Exporter).to receive(:new).with(exporter: exporter).and_call_original
+      get :diabetes_monthly_district_data,
+        params: {id: region.slug, report_scope: "district", format: "csv", period: period.attributes}
```
Comment (govindkrjoshi) [inline spec/controllers/reports/regions_controller_spec.rb:645]: You can also test that the download file name would be different

File: spec/controllers/reports/regions_controller_spec.rb [hunk 10]
```diff
@@ -550,37 +573,163 @@
...
+    it "returns 401 when user is not authorized" do
+      facility
+
+      get :hypertension_monthly_state_data, params: {id: region.slug, report_scope: "state", format: "csv"}
+      expect(response.status).to eq(401)
+    end
+
+    it "returns 302 found with invalid region" do
+      facility
+      sign_in(cvho.email_authentication)
```
Comment (govindkrjoshi) [inline spec/controllers/reports/regions_controller_spec.rb:663]: [Style] You could wrap all the authenticated blocks in a single context, and `sign_in` in a `before` block.

File: spec/exporters/monthly_district_data/diabetes_spec.rb [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+        facility_size_rows = data_service.facility_size_rows
+        expect(facility_size_rows.count).to eq(1)
+        expect(facility_size_rows.first&.count).to eq(44)
+        expect(facility_size_rows).to eq(expected_facility_size_rows)
+      end
+    end
+
+    describe "#facility_rows" do
+      it "provides accurate numbers for individual facilities" do
+        missed_visit_patient
```
Comment (govindkrjoshi) [inline spec/exporters/monthly_district_data/diabetes_spec.rb:268]: [Style] Here as well, you can use a before block create patients instead of using let and forcing the creation by referencing them.

File: spec/exporters/monthly_district_data/exporter_spec.rb [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+  let(:organization) { FactoryBot.create(:organization) }
+  let(:facility_group) { create(:facility_group, organization: organization) }
+  let(:facility1) { create(:facility, name: "Facility 1", block: "Block 1 - alphabetically first", facility_group: facility_group, facility_size: :community) }
+  let(:facility2) { create(:facility, name: "Facility 2", block: "Block 2 - alphabetically second", facility_group: facility_group, facility_size: :community) }
+  let(:region) { facility1.region.district_region }
+  let(:period) { Period.month(Date.today) }
+
+  describe "#report" do
+    before do
+      facility1
```
Comment (govindkrjoshi) [inline spec/exporters/monthly_district_data/exporter_spec.rb:22]: [Style] Same here

File: app/controllers/reports/regions_controller.rb (hunk 4 above)
Comment (PriyangaPKini) [reply app/controllers/reports/regions_controller.rb:312]: yeah, we should

File: app/models/reports/region_summary_schema.rb (hunk 5 above)
Comment (PriyangaPKini) [reply app/models/reports/region_summary_schema.rb:355]: I think we shouldn't because we don't prefix `adjusted_` to any of the `region_summary_schema` methods

File: spec/controllers/reports/regions_controller_spec.rb (hunk 8 above)
Comment (PriyangaPKini) [reply spec/controllers/reports/regions_controller_spec.rb:595]: Right, will make the changes

File: spec/controllers/reports/regions_controller_spec.rb (hunk 9 above)
Comment (PriyangaPKini) [reply spec/controllers/reports/regions_controller_spec.rb:645]: Done

File: spec/controllers/reports/regions_controller_spec.rb (hunk 10 above)
Comment (PriyangaPKini) [reply spec/controllers/reports/regions_controller_spec.rb:663]: done

File: app/models/reports/region_summary_schema.rb (hunk 5 above)
Comment (govindkrjoshi) [reply app/models/reports/region_summary_schema.rb:355]: But there is  specific meaning to `adjusted_`, i.e., patients registered 3 or more months earlier. 
If `diabetes_patients_with_bs_taken` is a count of patients registered 3 or more months earlier, it should be prefixed with `adjusted_`

File: app/components/dashboard/diabetes/downloads_dropdown_component.html.erb [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+        <% end %>
+      <% end %>
+    <% end %>
+
+    <% unless region.region_type.in?(%w[block]) %>
+     <div class="dropdown-divider"></div>
+    <% end %>
+
+    <h6 class="dropdown-header c-grey-dark">
+      <%= "#{quarter_string(Date.current.prev_quarter)}" %> Snapshots
```
Comment (govindkrjoshi) [inline app/components/dashboard/diabetes/downloads_dropdown_component.html.erb:51]: We should hide this section since it's not implemented yet. 

Also, we should check if we need to implement this section at all.

File: spec/exporters/monthly_district_data/diabetes_spec.rb (hunk 11 above)
Comment (PriyangaPKini) [reply spec/exporters/monthly_district_data/diabetes_spec.rb]: done

File: spec/exporters/monthly_district_data/exporter_spec.rb (hunk 12 above)
Comment (PriyangaPKini) [reply spec/exporters/monthly_district_data/exporter_spec.rb]: done

File: app/models/reports/region_summary_schema.rb [hunk 14]
```diff
@@ -350,9 +354,9 @@ def visited_without_bp_taken(with_ltfu: false)
...
       region_period_cached_query(__method__) do |entry|
-        diabetes_under_care(:bs_below_200)[entry.region.slug][entry.period] +
-          diabetes_under_care(:bs_200_to_300)[entry.region.slug][entry.period] +
-          diabetes_under_care(:bs_over_300)[entry.region.slug][entry.period]
+        adjusted_diabetes_under_care(:bs_below_200)[entry.region.slug][entry.period] +
+          adjusted_diabetes_under_care(:bs_200_to_300)[entry.region.slug][entry.period] +
+          adjusted_diabetes_under_care(:bs_over_300)[entry.region.slug][entry.period]
```
Comment (PriyangaPKini) [inline app/models/reports/region_summary_schema.rb:359]: Consumers who use `region_summary_schema` need `adjusted_` data from the table but do not want to mention it every time they use it. Hence we are not prefixing the methods with `adjusted_`. Only the ` adjusted_diabetes_under_care` method is out of this convention.

File: app/models/reports/region_summary_schema.rb [hunk 15]
```diff
@@ -389,21 +393,21 @@ def visited_without_bp_taken(with_ltfu: false)
...
       region_period_cached_query(__method__, blood_sugar_risk_state: blood_sugar_risk_state) do |entry|
         rounded_percentages({
-          random: diabetes_under_care(blood_sugar_risk_state, :random)[entry.region.slug][entry.period],
-          post_prandial: diabetes_under_care(blood_sugar_risk_state, :post_prandial)[entry.region.slug][entry.period],
-          fasting: diabetes_under_care(blood_sugar_risk_state, :fasting)[entry.region.slug][entry.period],
-          hba1c: diabetes_under_care(blood_sugar_risk_state, :hba1c) [entry.region.slug][entry.period]
+          random: adjusted_diabetes_under_care(blood_sugar_risk_state, :random)[entry.region.slug][entry.period],
+          post_prandial: adjusted_diabetes_under_care(blood_sugar_risk_state, :post_prandial)[entry.region.slug][entry.period],
+          fasting: adjusted_diabetes_under_care(blood_sugar_risk_state, :fasting)[entry.region.slug][entry.period],
+          hba1c: adjusted_diabetes_under_care(blood_sugar_risk_state, :hba1c) [entry.region.slug][entry.period]
```
Comment (PriyangaPKini) [inline app/models/reports/region_summary_schema.rb:399]: https://github.com/simpledotorg/simple-server/pull/4114/files#r925338357

File: app/models/reports/region_summary_schema.rb [hunk 16]
```diff
@@ -389,21 +393,21 @@ def visited_without_bp_taken(with_ltfu: false)
...
       region_period_cached_query(__method__, blood_sugar_risk_state: blood_sugar_risk_state) do |entry|
         {
-          random: diabetes_under_care(blood_sugar_risk_state, :random)[entry.region.slug][entry.period],
-          post_prandial: diabetes_under_care(blood_sugar_risk_state, :post_prandial)[entry.region.slug][entry.period],
-          fasting: diabetes_under_care(blood_sugar_risk_state, :fasting)[entry.region.slug][entry.period],
-          hba1c: diabetes_under_care(blood_sugar_risk_state, :hba1c) [entry.region.slug][entry.period]
+          random: adjusted_diabetes_under_care(blood_sugar_risk_state, :random)[entry.region.slug][entry.period],
+          post_prandial: adjusted_diabetes_under_care(blood_sugar_risk_state, :post_prandial)[entry.region.slug][entry.period],
+          fasting: adjusted_diabetes_under_care(blood_sugar_risk_state, :fasting)[entry.region.slug][entry.period],
+          hba1c: adjusted_diabetes_under_care(blood_sugar_risk_state, :hba1c) [entry.region.slug][entry.period]
```
Comment (PriyangaPKini) [inline app/models/reports/region_summary_schema.rb:410]: https://github.com/simpledotorg/simple-server/pull/4114/files#r925338357

File: spec/exporters/monthly_district_report/diabetes/block_data_spec.rb [hunk 17]
```diff
@@ -0,0 +1,136 @@
+require "rails_helper"
+
+def mock_block_repo(repo, district, month)
```
Comment (PriyangaPKini) [inline spec/exporters/monthly_district_report/diabetes/block_data_spec.rb:3]: Since this class directly calls the repo to get the data, it doesn't make sense to mock the repo here.
Also, even if the structure of the repo changes the tests will continue to pass.

File: spec/exporters/monthly_district_report/diabetes/district_data_spec.rb [hunk 18]
```diff
@@ -0,0 +1,175 @@
+require "rails_helper"
+
+def mock_district_repo(repo, district, month)
```
Comment (PriyangaPKini) [inline spec/exporters/monthly_district_report/diabetes/district_data_spec.rb:3]: https://github.com/simpledotorg/simple-server/pull/4114/files#r925405100

File: spec/exporters/monthly_district_report/diabetes/facility_data_spec.rb [hunk 19]
```diff
@@ -0,0 +1,132 @@
+require "rails_helper"
+
+def mock_facility_repo(repo, district, month)
```
Comment (PriyangaPKini) [inline spec/exporters/monthly_district_report/diabetes/facility_data_spec.rb:3]: https://github.com/simpledotorg/simple-server/pull/4114/files#r925405100

File: spec/exporters/monthly_district_data/diabetes_spec.rb [hunk 20]
```diff
@@ -0,0 +1,249 @@
...
+  around do |example|
+    # This is in the style of ReportingHelpers::freeze_time_for_reporting_specs.
+    # Since FacilityAppointmentScheduledDays only keeps the last 6 months of data, the date cannot be a
+    # fixed point in time like the spec helper.
+    Timecop.freeze("#{Date.today.end_of_month} 23:00 IST") do
+      example.run
+    end
+  end
+  before do
+    @organization = FactoryBot.create(:organization)
```
Comment (prabhanshuguptagit) [inline spec/exporters/monthly_district_data/diabetes_spec.rb:13]: Prefer using a method to setup data instead of instance variables.

File: spec/exporters/monthly_state_data/diabetes_spec.rb [hunk 21]
```diff
@@ -0,0 +1,246 @@
...
+  around do |example|
+    # This is in the style of ReportingHelpers::freeze_time_for_reporting_specs.
+    # Since FacilityAppointmentScheduledDays only keeps the last 6 months of data, the date cannot be a
+    # fixed point in time like the spec helper.
+    Timecop.freeze("#{Date.today.end_of_month} 23:00 IST") do
+      example.run
+    end
+  end
+
+  let(:organization) { FactoryBot.create(:organization) }
```
Comment (PriyangaPKini) [inline spec/exporters/monthly_state_data/diabetes_spec.rb:13]: https://github.com/simpledotorg/simple-server/pull/4114/files#r925405328

File: spec/exporters/monthly_state_data/exporter_spec.rb [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+  around do |example|
+    # This is in the style of ReportingHelpers::freeze_time_for_reporting_specs.
+    # Since FacilityAppointmentScheduledDays only keeps the last 6 months of data, the date cannot be a
+    # fixed point in time like the spec helper.
+    Timecop.freeze("#{Date.today.end_of_month} 23:00 IST") do
+      example.run
+    end
+  end
+
+  let(:organization) { FactoryBot.create(:organization) }
```
Comment (PriyangaPKini) [inline spec/exporters/monthly_state_data/exporter_spec.rb:13]: https://github.com/simpledotorg/simple-server/pull/4114/files#r925405328

File: spec/exporters/monthly_state_data/hypertension_spec.rb [hunk 23]
```diff
@@ -0,0 +1,242 @@
...
+  around do |example|
+    # This is in the style of ReportingHelpers::freeze_time_for_reporting_specs.
+    # Since FacilityAppointmentScheduledDays only keeps the last 6 months of data, the date cannot be a
+    # fixed point in time like the spec helper.
+    Timecop.freeze("#{Date.today.end_of_month} 23:00 IST") do
+      example.run
+    end
+  end
+
+  let(:organization) { FactoryBot.create(:organization) }
```
Comment (PriyangaPKini) [inline spec/exporters/monthly_state_data/hypertension_spec.rb:13]: https://github.com/simpledotorg/simple-server/pull/4114/files#r925405328

File: app/exporters/patients_with_history_exporter.rb [hunk 24]
```diff
@@ -156,28 +153,12 @@ def csv_fields(patient_summary)
...
-      end,
-      latest_blood_sugar_date(patient_summary),
-      patient_summary.latest_blood_sugar.to_s,
-      latest_blood_sugar_type(patient_summary)
+      (1..display_blood_pressures).map { |i| bp_fields(latest_bps, i, patient_appointments: patient_appointments, all_medications: all_medications) },
+      display_blood_sugars == 0 ?
+        [latest_blood_sugar_date(patient_summary),
+          patient_summary.latest_blood_sugar.to_s,
+          latest_blood_sugar_type(patient_summary)] :
+        (1..display_blood_sugars).map { |i| blood_sugar_fields(latest_blood_sugars, i, patient_appointments: patient_appointments) }
```
Comment (prabhanshuguptagit) [inline app/exporters/patients_with_history_exporter.rb:161]: This ternary is hard to parse, consider replacing with an if.

File: app/controllers/reports/regions_controller.rb [hunk 25]
```diff
@@ -272,22 +280,72 @@ def monthly_state_data_report
...
+
+  def diabetes_monthly_district_report
+    return unless current_admin.feature_enabled?(:monthly_district_report)
+
+    monthly_district_report(
+      MonthlyDistrictReport::Exporter.new(
+        facility_data: MonthlyDistrictReport::Diabetes::FacilityData.new(@region, @period),
+        block_data: MonthlyDistrictReport::Diabetes::BlockData.new(@region, @period),
+        district_data: MonthlyDistrictReport::Diabetes::DistrictData.new(@region, @period)
+      ),
```
Comment (prabhanshuguptagit) [inline app/controllers/reports/regions_controller.rb:342]: [suggestion] We spoke about sub-classing these as separate exporter classes from an `Exporter` instead of composition-ing.
**pros**
- the contract between the exporter and the "data" classes is more explicit.
- can share code between `MonthlyDistrictReport::Diabetes` and `MonthlyDistrictReport::Hypertension`

**cons**
- inheriting all this functionality in couples exporting and fetching data from repository

**Review by prabhanshuguptagit — APPROVED** [review]
